posts/2024-06-13_reaching_consensus/README.md

```diff
@@ -0,0 +1,475 @@
+
+We call this tuple denoting the current consensus height and round `(h, r)` a ***view***.
+
+For leader-based consensus algorithms like Tendermint, one node (validator) called the ***proposer*** (who is deterministically selected by all participating nodes) needs to propose a value for the current consensus view. This ***proposal*** value can in theory be anything, but in the context of blockchain systems, this is usually a block containing transactions (state transitions).
+
+Specific to Tendermint, in order to reach consensus, there needs to be an agreement between at least `2F+1` validator nodes (where `F` denotes the maximum number of faulty nodes that are tolerated in the system). In the broader Tendermint ecosystem, validators are not equal, since the consensus protocol mostly used with Tendermint is Proof of Stake, and the validator’s power is proportional to their stake. In this case, `F` denotes the maximum voting power of faulty nodes that are tolerated in the system. In practice, `2F+1` is 2/3+ (supermajority) of the validator set’s voting power.
+
+## How does Tendermint work?
+
+The term “[Tendermint](https://docs.tendermint.com/v0.34/introduction/what-is-tendermint.html)” refers to a much larger scope of functionality than what we discuss in this article.  By “Tendermint” here, we solely refer to the Tendermint consensus engine (state machine).
+
+A good graphic that shows an overview of the Tendermint state machine can be found on the official [Tendermint Core documentation](https://docs.tendermint.com/v0.34/introduction/what-is-tendermint.html#consensus-overview):
+
+[![polka](https://gnolang.github.io/blog/2024-06-13_reaching_consensus/src/thumbs/polka.png)](https://gnolang.github.io/blog/2024-06-13_reaching_consensus/src/polka.png)
+
+However, we want to tackle this question of Tendermint’s inner-workings by weaving in the theory from the paper titled *The latest gossip on BFT consensus*, and applying it in practice.
+
+On a surface-level, the Tendermint consensus engine has 3 states:
+- **Propose** - the *proposer* for the current view (height, round) sends out a *proposal* to all other validator nodes. Non-proposers wait for the proposal and verify it
+- **Prevote** - the validator nodes wait for 2F+1 `PREVOTE` messages, potentially locking on a proposal
+- **Precommit** - the validator nodes wait for 2F+1 `PRECOMMIT` messages, potentially saving the block to the block store, and starting a new consensus round (with a view `(h+1, 0)`)
+
+These states are not as straightforward as they’re made out to be — there is an added complication of *state timers* which guarantee protocol liveness, and other asynchronous situations like consensus round jumps within the same height. We will discuss these mechanisms later on, as they are critical to making sure Tendermint is ticking correctly.
+
+### I. Propose
```

Comment:
Roman numerals should not have a period after them. I suggest just switching it out to Arabic numerals and keep the period.